server/src/main.rs

```diff
@@ -304,13 +306,43 @@ async fn serve(interface: &str, conf: &ServerConfig) -> Result<(), Error> {
 
     log::info!("innernet-server {} starting.", VERSION);
     let routes = routes(context.clone()).with(warp::log("warp")).boxed();
-    warp::serve(routes)
-        .run((config.address, config.listen_port))
-        .await;
+
+    let listener = get_listener((config.address, config.listen_port).into(), interface)?;
+
+    let warp_svc = warp::service(routes);
+    let make_svc = hyper::service::make_service_fn(move |socket: &AddrStream| {
```

Comment:
this is necessary because warp does not expose a way to use a user-specified `TcpListener`, I suppose?

Reply:
Yep, that's right - I also had to fork warp for now, pending https://github.com/seanmonstar/warp/issues/830.